python-stdlib/time/time.py

```diff
@@ -0,0 +1,54 @@
+import re
+
+_TS_YEAR = const(0)
+_TS_MON = const(1)
+_TS_MDAY = const(2)
+_TS_HOUR = const(3)
+_TS_MIN = const(4)
+_TS_SEC = const(5)
+_TS_WDAY = const(6)
+_TS_YDAY = const(7)
+_TS_ISDST = const(8)
+
+_WDAY = const(("Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"))
+_MDAY = const(
+    (
+        "January",
+        "February",
+        "March",
+        "April",
+        "May",
+        "June",
+        "July",
+        "August",
+        "September",
+        "October",
+        "November",
+        "December",
+    )
+)
+
+_time_regex = None
+
+
+def strftime(datefmt, ts):
+    global _time_regex
+    if _time_regex is None:
+        _time_regex = re.compile(r"%(-?[a-zA-Z])")
+    return _time_regex.sub(r"%(\1)s", datefmt) % {
+        "a": _WDAY[ts[_TS_WDAY] - 1][0:3],
+        "A": _WDAY[ts[_TS_WDAY] - 1],
+        "b": _MDAY[ts[_TS_MON] - 1][0:3],
+        "B": _MDAY[ts[_TS_MON] - 1],
+        "d": f"{ts[_TS_MDAY]:02d}",
+        "H": f"{ts[_TS_HOUR]:02d}",
+        "I": f"{ts[_TS_HOUR]%12:02d}",
+        "j": f"{ts[_TS_YDAY]:03d}",
+        "m": f"{ts[_TS_MON]:02d}",
+        "M": f"{ts[_TS_MIN]:02d}",
+        "P": "AM" if ts[_TS_HOUR] < 12 else "PM",
+        "S": f"{ts[_TS_SEC]:02d}",
+        "w": f"{ts[_TS_WDAY]}",
+        "y": f"{ts[_TS_YEAR]%100:02d}",
+        "Y": f"{ts[_TS_YEAR]}",
+    }
```

Comment:
This is a neat way to implement strftime (using the dict and `%` formatting), but it will contribute to a lot of heap allocations, and most of the strings created here will never be used by the caller of `strftime`.

For example, if logging happens frequently (eg a few times per second) then this is going to put a lot of pressure on the GC heap.  Each call to `strftime` creates about 12 strings, only one of which is the actual formatted string.

I don't really have a good alternative, except to implement this as a string parser that looks for the `%` patterns and substitutes them on demand.  Ie `for c in datefmt: ...`.

Reply:
I do prefer the new `strftime2` version, simply because it will perform less allocations (and may be faster overall due to less heap allocations).

But I suggest using `io.String()` to build the output string, so it doesn't need to create a new string for each `+=`.  Also please do `import io` within the function so that ports that don't have it can still load this code (and just not use `strftime`).

Furthermore, I think the code can be a little simpler with this logic:
```py
for k in datefmt:
    if fmtsp:
        if k == "a":
            ..
        else:
           ftime += k
        fmtsp = False
    elif k == "%":
        fmtsp = True
    else:
        ftime += k
```

Reply:
> within the function so that ports that don't have it can still load this code (and just not use strftime).

@dpgeorge Can we use a bytearray maybe so it's available on most (all?) ports ?

Reply:
Or a list and then `"".join(ftime))`

Reply:
A bytearray might have problems with unicode characters (or it may be hard to handle them properly).  Using list and join will use a bit of extra memory for the list itself.

I think `io.StringIO()` is the best way to do it.  And from `py/mpconfig.h`:
```c
#define MICROPY_PY_BUILTINS_STR_OP_MODULO (MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_CORE_FEATURES)
#define MICROPY_PY_BUILTINS_BYTEARRAY (MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_CORE_FEATURES)
#define MICROPY_PY_IO (MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_CORE_FEATURES)
```
That means that, %-formatting, bytearray and the `io` module are all available at the same feature level.  So `io.StringIO()` will most likely be available on ports that use this function (as long as they use the feature level config, which we are trying to make sure all ports do).

Reply:
> Using list and join will use a bit of extra memory for the list itself.

Changed to StringIO.